### PR TITLE
Optimizations for multiple quadword precision integer support.

### PR DESCRIPTION
**Pull request by munroesj52**
Added multiply-add operations to halfword (POWER7) and word (POWER8)
in support of multiple quadword multiply.
- Used \copybrief for halfword and word operations, that are implemented in other headers, so they appear in alphabetical order with other halfword and word operations.
- Added appropriate unit and compile tests.

Changelog:

	* doc/pveclib-doxygen-pveclib.doxy [PAPER_TYPE]:
	Change to letter.
	* doc/pveclibmaindox.h [mainpage_ref_docs]: Rework references
	to avoid line overflows in generated PDF.

	* src/pveclib/vec_bcd_ppc.h (vec_bcdcfsq): Doxygen edits.
	(vec_unpack_Decimal128):  Doxygen edits.
	* src/pveclib/vec_int16_ppc.h: General Doxygen edits.
	(i16_recent_additions): new doxygen section.
	(vec_vmaddeuh, vec_vmaddouh): New operation.
	* src/pveclib/vec_int32_ppc.h
	(i16_recent_additions): new doxygen section.
	(vec_vmuleuw, vec_vmulouw): Add forward declare.
	(vec_mrgahw, vec_mrgalw): Add doxygen note.
	(vec_muleuw, vec_mulouw): Simplify implementation using
	vec_vmuleuw, vec_vmulouw.
	(vec_vmadd2euw, vec_vmadd2ouw, vec_vmaddeuw, vec_vmaddouw,
	vec_vmsumuwm): Doxygen copybrief from vec_int64_ppc.h.
	(vec_vmuleuw, vec_vmulouw): New operations.

	* src/testsuite/arith128_test_i16.c (test_vmadduh):
	New Unit test.
	(test_vec_i16): Add test_vmadduh to test driver.

	* src/testsuite/vec_int16_dummy.c
	(__test_vmaddeuh, __test_vmaddouh, __test_vmaddouh_alt,
	__test_vadduqm_alt7, __test_vaddcuq_alt7, __test_vmadduh_alt7):
	New compile tests.
	* src/testsuite/vec_int32_dummy.c
	(__test_vmuleuw, __test_vmulouw): New compile tests.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

### CODE REVIEW

File: src/pveclib/vec_int32_ppc.h [hunk 1]
```diff
@@ -1419,4 +1300,257 @@ vec_srwi (vui32_t vra, const unsigned int shb)
...
+ * a final vadduwm to complete the carry propagation for the
+ * doubleword add.
+ *
+ * \note This implementation is NOT endian sensitive and the function
+ * is stable across BE/LE implementations.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   |   7   | 2/cycle  |
+ * |power9   |   7   | 2/cycle  |
```
Comment (ThinkOpenly) [inline src/pveclib/vec_int32_ppc.h]: No power7 data? (I doubt there's too much interest, but you did go to a lot of trouble to support POWER7.

File: src/pveclib/vec_int32_ppc.h [hunk 2]
```diff
@@ -1419,4 +1300,257 @@ vec_srwi (vui32_t vra, const unsigned int shb)
...
+    c = vec_sld (c, zero, 4);
+    res = (vui64_t)vec_vadduwm (r, c);
+  }
+#endif
+  return (res);
+}
+
+/** \brief Vector Multiply Odd Unsigned Words.
+ *
+ * Multiple the odd words of two vector unsigned int values and return
```
Comment (ThinkOpenly) [inline src/pveclib/vec_int32_ppc.h]: s/Multiple/Multiply/
Also, same comments with respect to bit size of types.
So, these are now global comments.

**Comment by ThinkOpenly** [comment]
> Added multiply-add operations to halfword (POWER7) and word (POWER8)
in support of multiply quadword multiply.

"multiply quadword multiply" ?

> Used copybrief to that halfword and word operations, that are
implemented in other headers, appear in aphabetical order with other
halfword and word operations.

- "Used copybrief to that halfword and word operations" doesn't read well.  "Used `\copybrief` to copy the briefs from those halfword and word operations that are implemented in other headers" ?
- s/aphabetical/alphabetical/

still reviewing the code...

**Comment by ThinkOpenly** [comment]
Also, your commit subject seems to have "integersupport". (needs a space)

File: src/pveclib/vec_int32_ppc.h (hunk 1 above)
Comment (munroesj52) [reply src/pveclib/vec_int32_ppc.h]: There problem is there is no "public" version of the POWER7 book 4 (AKA users manual). These do exist for the POWER8/9 on OpenPOWER.org.

Before I retired I had access to the IBM confidential Book 4s, but i purged those during my last week at IBM.

I don't want to publish guesses and so choose to not publish any cycle times for P7.

File: src/pveclib/vec_int32_ppc.h [hunk 3]
```diff
@@ -1419,4 +1300,257 @@ vec_srwi (vui32_t vra, const unsigned int shb)
...
+ * \note this implementation exists in
+ * \ref vec_int64_ppc.h::vec_vmsumuwm()
+ * as it requires vec_addudm().
+ */
+static inline vui64_t
+vec_vmsumuwm (vui32_t a, vui32_t b, vui64_t c);
+
+/** \brief Vector Multiply Even Unsigned words.
+ *
+ * Multiple the even words of two vector unsigned int values and return
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/pveclib/vec_int32_ppc.h]: auto correct my nemesis, Ok

On the wording of element size I have been back and forth. I think in this case I was seeing problems with line overflow in the generated PDF for @param. So decided to shorten to text "vector [unsigned | signed] [char | short | int | long]" which is unambiguous for PPC64.

For now please open a new issue for "should a consistent wording for element type/size for vector @param"

File: src/pveclib/vec_int32_ppc.h (hunk 2 above)
Comment (munroesj52) [reply src/pveclib/vec_int32_ppc.h]: Ok, see above.

**Comment by munroesj52** [comment]
Thanks Paul. I think should work on consistent language for descriptions after we get past this bow wave of changes

**Pull request merged**